### PR TITLE
some small but useful changes

### PR DESCRIPTION
**Pull request by a8156268**

### CODE REVIEW

File: bt-core/src/main/java/bt/net/pipeline/SocketChannelHandler.java [hunk 1]
```diff
@@ -139,12 +139,18 @@ public void flush() {
...
                 // buffer has been released
+                outboundBuffer.unlock();
                 return;
             }
             buffer.flip();
             try {
                 while (buffer.hasRemaining()) {
-                    channel.write(buffer);
+                    int written = channel.write(buffer);
+                    // for unknown reason, -2 is returned of every call to channel.write, and no data is written, buffer.hasRemaining() is always true.
```
Comment (atomashpolskiy) [inline bt-core/src/main/java/bt/net/pipeline/SocketChannelHandler.java:149]: `SocketChannel.write` returns `-2`??

File: bt-core/src/main/java/bt/torrent/data/DefaultDataWorker.java [hunk 2]
```diff
@@ -58,15 +60,17 @@ public DefaultDataWorker(IRuntimeLifecycleBinder lifecycleBinder,
         this.verifier = verifier;
         this.blockCache = blockCache;
 
-        this.executor = Executors.newSingleThreadExecutor(new ThreadFactory() {
+        this.executor = new ThreadPoolExecutor(config.getDataWorkerCorePoolSize(),
```
Comment (atomashpolskiy) [inline bt-core/src/main/java/bt/torrent/data/DefaultDataWorker.java:63]: Have you verified that the futures are thread-safe? I'm trying to recall why I ended up having a single-threaded data worker, but can't..

File: bt-core/src/main/java/bt/torrent/messaging/TorrentWorker.java [hunk 3]
```diff
@@ -105,6 +110,47 @@ public TorrentWorker(TorrentId torrentId,
         eventSource.onPeerConnected(torrentId, e -> onPeerConnected(e.getConnectionKey()));
 
         eventSource.onPeerDisconnected(torrentId, e -> onPeerDisconnected(e.getConnectionKey()));
+
+        startAssignmentsUpdateThread(torrentId, eventSource, config);
+    }
+
+    private void startAssignmentsUpdateThread(TorrentId torrentId, EventSource eventSource, Config config) {
```
Comment (atomashpolskiy) [inline bt-core/src/main/java/bt/torrent/messaging/TorrentWorker.java:117]: Same as for DataWorker, I'm concerned about potential concurrency issues (with regards to the objects that are updated here and then used in the other parts of the program).

**Comment by atomashpolskiy** [comment]
Awesome updates as always, but I'm puzzled by the `SocketChannel.write` comment and also feel a bit uncomfortable about multithreading stuff.

File: bt-core/src/main/java/bt/net/pipeline/SocketChannelHandler.java (hunk 1 above)
Comment (wizzardo) [reply bt-core/src/main/java/bt/net/pipeline/SocketChannelHandler.java:149]: not sure about -2, but I also have an issue here, sometimes connection just stuck in infinite loop trying to write some messages, I've tried to add `key.interestOps(SelectionKey.OP_WRITE)`, so for now I fixed it for myself with just closing the connection as soon as I get `written == 0`, at least it doesn't consume cpu in the loop

**Comment by atomashpolskiy** [comment]
Yeah, I also observed this and spent quite some time trying to diagnose and
debug the root cause, but failed. Frankly speaking, network code (first and
foremost, NIO loop) is pretty bad and not thought through properly and
suffers heavily from lack of KISS.

File: bt-core/src/main/java/bt/net/pipeline/SocketChannelHandler.java (hunk 1 above)
Comment (a8156268) [reply bt-core/src/main/java/bt/net/pipeline/SocketChannelHandler.java:149]: I tried to reproduce this problem, but failed.    
But this -2 infinite loop did happen in my previous tests, and this changes fixed it.    
Any way, negative returned value is always bad, and we should handle this.(Throw exception looks like a reasonable operation)

Some hints about negative values: 
In this file sun.nio.ch.IOStatus, some constants is defined:
image

File: bt-core/src/main/java/bt/torrent/data/DefaultDataWorker.java (hunk 2 above)
Comment (a8156268) [reply bt-core/src/main/java/bt/torrent/data/DefaultDataWorker.java:63]: I am not 100% sure they are thread-safe.  
I just did some tests about my own using cases, though these test passes, but they may not enough.   
I will do more tests later and recommit again.

File: bt-core/src/main/java/bt/net/pipeline/SocketChannelHandler.java (hunk 1 above)
Comment (wizzardo) [reply bt-core/src/main/java/bt/net/pipeline/SocketChannelHandler.java:149]: what jvm do you use? In openjdk returned status is always normalized through 
```
//sun.nio.ch.IOStatus
    public static int normalize(int var0) {
        return var0 == -2 ? 0 : var0;
    }
```
so, actually -2 means 0, and it's ok for non-blocking sockets, output buffer is full and we should wait. But I still had the situation when OP_WRITE didn't happen for the socket

File: bt-core/src/main/java/bt/torrent/data/DefaultDataWorker.java (hunk 2 above)
Comment (atomashpolskiy) [reply bt-core/src/main/java/bt/torrent/data/DefaultDataWorker.java:63]: What platform are you testing on? (x86, arm, ...)

File: bt-core/src/main/java/bt/net/pipeline/SocketChannelHandler.java (hunk 1 above)
Comment (a8156268) [reply bt-core/src/main/java/bt/net/pipeline/SocketChannelHandler.java:149]: here is the jdk version
java version "1.8.0_60"
Java(TM) SE Runtime Environment (build 1.8.0_60-b27)
Java HotSpot(TM) 64-Bit Server VM (build 25.60-b23, mixed mode)

Comment (a8156268) [reply bt-core/src/main/java/bt/net/pipeline/SocketChannelHandler.java:149]: It's a mistake.
Maybe is observed a -2 is returned before normalize,but incorrectly recognize this value as returned by write.

File: bt-core/src/main/java/bt/torrent/data/DefaultDataWorker.java (hunk 2 above)
Comment (a8156268) [reply bt-core/src/main/java/bt/torrent/data/DefaultDataWorker.java:63]: x86

**Pull request closed without merging**